nexus/src/app/instance.rs

```diff
@@ -654,33 +654,24 @@ impl super::Nexus {
         instance_name: &Name,
         params: &params::NetworkInterfaceCreate,
     ) -> CreateResult<db::model::NetworkInterface> {
-        let (.., authz_project, authz_instance, db_instance) =
+        let (.., authz_project, authz_instance) =
             LookupPath::new(opctx, &self.db_datastore)
                 .organization_name(organization_name)
                 .project_name(project_name)
                 .instance_name(instance_name)
-                .fetch()
+                .lookup_for(authz::Action::Modify)
                 .await?;
 
-        // TODO-completeness: We'd like to relax this once hot-plug is
-        // supported.
-        //
-        // TODO-correctness: There's a TOCTOU race here. Someone might start the
-        // instance between this check and when we actually create the NIC
-        // record. One solution is to place the state verification in the query
-        // to create the NIC. Unfortunately, that query is already very
-        // complicated. See
-        // https://github.com/oxidecomputer/omicron/issues/1134.
-        let stopped =
-            db::model::InstanceState::new(external::InstanceState::Stopped);
-        if db_instance.runtime_state.state != stopped {
-            return Err(external::Error::invalid_request(
-                "Instance must be stopped to attach a new network interface",
-            ));
-        }
-
         // NOTE: We need to lookup the VPC and VPC Subnet, since we need both
         // IDs for creating the network interface.
+        //
+        // TODO-correctness: There are additional races here. The VPC and VPC
```

Comment:
> If it isn't one, it's another, eh?

Indeed :)

>  I really want better testing / analysis to help us suss these issues out

Yeah, that'd be nice. I'm not sure how to force ourselves to hit this, since the flow would be: fetch the VPC Subnet successfully, delete the VPC Subnet, then try to run this query. How would we do that? Maybe part of a set of stress tests?

Reply:
I'm not sure. I can think of a view options:

- Random-walk stress testing: ideally, having a "set of operations" and "set of objects/names" that can all be acted upon, and doing some random-walk style operations would be neat.
- Permutation testing: we could try to integrate something like https://github.com/tokio-rs/loom to invasively test the DB-related ops. There would still be reliance on CRDB, however, so these tests would probably not run as a part of the CI check - they might be delegated to "once daily" jobs.
- Model checking: we could *attempt* to model the DB interactions externally (e.g., through TLA+) and try to prove safety properties about them. I'm skeptical about the cost/benefit of this approach; it seems like a more expensive option.

Reply:
I like the second approach. The first is probably easiest, but I'd be worried about it building our assumptions about how things _should_ be used. I agree the formal verification approach is likely to be too heavyweight.